cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/StatefulSetOperator.java

```diff
@@ -205,15 +208,53 @@ public void onClose(KubernetesClientException e) {
         }
     }
 
+    private static ObjectMeta templateMetadata(StatefulSet resource) {
+        return resource.getSpec().getTemplate().getMetadata();
+    }
+
     public String getPodName(StatefulSet desired, int podId) {
-        return desired.getMetadata().getName() + "-" + podId;
+        return templateMetadata(desired).getName() + "-" + podId;
+    }
+
+    private void setGeneration(StatefulSet desired, int nextGeneration) {
+        templateMetadata(desired).getAnnotations().put(ANNOTATION_GENERATION, String.valueOf(nextGeneration));
+    }
+
+    private static int getGeneration(ObjectMeta objectMeta) {
+        if (objectMeta.getAnnotations().get(ANNOTATION_GENERATION) == null) {
+            new RuntimeException().printStackTrace();
+        }
+        return Integer.parseInt(objectMeta.getAnnotations().getOrDefault(ANNOTATION_GENERATION, "-2"));
+    }
+
+    protected void incrementGeneration(StatefulSet current, StatefulSet desired) {
+        final int generation = Integer.parseInt(templateMetadata(current).getAnnotations().getOrDefault(ANNOTATION_GENERATION, "1"));
+        final int nextGeneration = generation + 1;
+        setGeneration(desired, nextGeneration);
+    }
+
+    protected abstract boolean shouldIncrementGeneration(StatefulSet current, StatefulSet desired);
+
+    private static int getSsGeneration(StatefulSet resource) {
+        if (resource == null) {
+            return -3;
+        }
+        return getGeneration(templateMetadata(resource));
+    }
+
+    private static int getPodGeneration(Pod resource) {
+        if (resource == null) {
+            return -4;
+        }
+        return getGeneration(resource.getMetadata());
```

Comment:
Do these -3 and -4 values mean something useful?

Reply:
Ah thankyou, that's left over from debugging a weird problem I had with the tests, I'll fix this.